Improve explanation of why "the `ImmutableList.Builder` rule" produces `MINUS_NULL`. (#657)
jspecify/jspecify: docs/docs/spec.md

    > demonstrated in
    > https://github.com/jspecify/jspecify-reference-checker/pull/197.

    > The purpose of this part of the subsitution rules is ensure that non-null
    > types stay non-null during substitution, even if they don't have an
    > explicit `@NonNull` annotation on them.
    >
    > For an example of such a type, consider `Comparable`, a `@NullMarked`
    > interface that declares a non-nullable type parameter `T` and a method
    > `compare(T)`. By JSpecify rules, the method parameter has type `T
    > NO_CHANGE`, and that type is null-exclusive under every parameterization
    > in all worlds. Now consider a null-unmarked class that declares a method

>     > `Comparable<Foo> foo()`, which by JSpecify rules has a type argument `Foo

Reviewer: Is it worth having `@NonNull Comparable<Foo>` here, to avoid questions about dereferencing `foo()` below? Or is that too distracting, and people will know what you mean?

    > UNSPECIFIED`. In this example, the question is what type

>     > `foo().compare(...)` accepts, and the answer changes as a result of this

Reviewer: Framing this as "the answer changes" suggests dynamic rules somehow.

What about "this part of the substitution rule gives us a better answer"?

    > part of the subsitution rules:
    >

>     > -   If JSpecify were to directly subsitute `Foo UNSPECIFIED` in for `T`,
>     >     then the parameter type, which started out as non-null, would become

Reviewer: ```suggestion
    > -   Without this part of the rule, JSpecify would directly subsitute `Foo UNSPECIFIED` in for `T`;
    >     then the parameter type, which started out as non-null, would become
```

    >     unspecified during subsitution. As a result, lenient checkers would
    >     allow the call `foo().compare(null)`, since `Foo UNSPECIFIED` is
    >     [null-inclusive under every parameterization] in [some world].

>     > -   To avoid that, JSpecify recognizes that the parameter is non-null, and

Reviewer: ```suggestion
    > -   To avoid that, JSpecify uses this rule to recognize that the parameter is non-null, and
```

    >     it performs substitution as if the parameter type were `T MINUS_NULL`

>     >     instead of `T NO_CHANGE`.

Author: I was tempted to add here: "JSpecify treats those types equivalently before substitution, so it makes sense for JSpecify to ensure that that are equivalent after substitution, too."

I couldn't decide whether that would make things better or worse.

Reviewer: I had to think for a bit before I understood what you meant by "those types", so I don't think that sentence helps.

But maybe if you said more explicitly something like "JSpecify treats `T NO_CHANGE` just like `T MINUS_NULL` when `T` is non-nullable, and this rule preserves that equivalence through substitution, even when substituting `UNSPECIFIED` types."

If you do that, you should do it after these two bullets, not as part of the second bullet.

>     >     instead of `T NO_CHANGE`.

Reviewer: Should you describe what the `compare(...)` argument type becomes, in parallel to what you do in the previous bullet?

Author: Good point.

    > demonstrated in
    > https://github.com/jspecify/jspecify-reference-checker/pull/197.

> 

Reviewer: Section headers inside bullets inside non-normative blocks inside parentheses.... where does it end?

    > The primary purpose of this special case is to improve behavior in "the

>     > `ImmutableList.Builder` case": Because `ImmutableList.Builder.add` always

Reviewer: I'd love to avoid referring to a Guava class. Could we motivate this with a hypothetical class (or one in the JDK&mdash;what about `Comparable`?), to make it clear that this isn't really a special case? It's asymmetrical, but that's a function of the fact that `Foo!` is a strict subtype of `Foo?`.

But if we're going to refer to `ImmutableList.Builder`, shouldn't we at least link to it? It's not the case that everyone reading this spec will be intimately familiar with that API.

Author: I made an attempt to rewrite this to be about `Comparable` and to make clear that it's not a special case. PTAL.

    > in all worlds. Now consider a null-unmarked class that declares a method
    > `Comparable<Foo> foo()`, which by JSpecify rules has a type argument `Foo
    > UNSPECIFIED`. In this example, the question is what type
    > `foo().compare(...)` accepts. That question demonstrates the effect of
    > this part of the subsitution rule:
    >
    > -   Without this part of the rule, JSpecify would directly subsitute `Foo

>     >     UNSPECIFIED` in for `T`. Then the parameter type, which started out as

Reviewer: Do we need the "in" here? "substituted X in for Y" vs "substituted X for Y".

_This is the most consequential comment I've ever made._

Author: Done.

(I started wondering whether we should say "for `T NO_CHANGE`" or otherwise attempt to be more precise, but then I stuffed it back down.)

    > `builder.add(null)`, it would find that it does.
    >
    > To solve this, we need a special case for substitution for null-exclusive

>     > type parameters like the one on `ImmutableList.Builder`, one that avoids
>     > producing a type with nullness operator `UNSPECIFIED`. Since the special

Reviewer: The two "one"s confused me. How about this?

```suggestion
    > type parameters (like the one on `ImmutableList.Builder`) that doesn't
    > produce a type with nullness operator `UNSPECIFIED`. Since the special
```
